### PR TITLE
[FSDP] full_state_dict impl

### PR DESCRIPTION
**Pull request by rohan-varma**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #73324
* #73323

Implements `state_dict` and `load_state_dict` APIs for FSDP, with the following limitations:

1. Does not support `state_dict_device` (i.e. specifying which device params should be on) which fairscale does currently support
2. Does not yet support offload of state_dict onto CPU
3. Loads state_dict on all ranks currently. In the future we could add support for loading this on only rank 0, to avoid redundancy across ranks as usually only one rank is responsible for saving/loading the model. Along with (2) this would enable larger models to have state_dict called.

As discussed in FSDP checkpoint API proposal, `state_dict` will basically be a `full_state_dict` where full parameters are returned on all ranks. As a result this implies that the model must actually be able to fit on a single GPU.

Differential Revision: [D34433514](https://our.internmc.facebook.com/intern/diff/D34433514/)

### CODE REVIEW

File: test/distributed/fsdp/test_fsdp_state_dict.py [hunk 1]
```diff
@@ -58,14 +69,109 @@ class TestFSDPStateDict(FSDPTest):
...
+    @skip_if_lt_x_gpu(2)
+    @parametrize(
+        "cpu_offload",
+        [CPUOffload(offload_params=True), CPUOffload(offload_params=False)],
+    )
+    @parametrize("fp16", [True, False])
+    def test_basic_save_and_load_full_state_dict(self, cpu_offload, fp16):
+        """
+        Tests that we can save a full_state_dict and load it into a blank model
+        with various configs such as fp17 and cpu offload and parameters
```
Comment (fegin) [inline test/distributed/fsdp/test_fsdp_state_dict.py:96]: typo, fp16?

Comment (fegin) [inline test/distributed/fsdp/test_fsdp_state_dict.py:96]: typo, fp16?

File: test/distributed/fsdp/test_fsdp_state_dict.py [hunk 2]
```diff
@@ -58,14 +69,109 @@ class TestFSDPStateDict(FSDPTest):
...
+        match as expected.
+        """
+        for model_call in [
+            partial(self._get_simple_nested_model, cpu_offload=cpu_offload),
+            partial(self._get_simple_model, cpu_offload=cpu_offload),
+        ]:
+            model = model_call()
+            fsdp_state_dict = _get_state_dict(model, cpu_offload.offload_params, fp16)
+            if fp16:
+                # Verify fp17 is the type
```
Comment (fegin) [inline test/distributed/fsdp/test_fsdp_state_dict.py:106]: ditto

File: torch/distributed/fsdp/fully_sharded_data_parallel.py [hunk 3]
```diff
@@ -846,7 +862,17 @@ def state_dict(self, *args, **kwargs):
         if torch.cuda.is_available():
             torch.cuda.synchronize()
         if self._state_dict_type == StateDictType.FULL_STATE_DICT:
-            return super().state_dict(*args, **kwargs)
+            self._lazy_init()
```
Comment (fegin) [inline torch/distributed/fsdp/fully_sharded_data_parallel.py:865]: Can we move this outside of the `if` to ensure all different state_dict call `_lazy_init`?

File: torch/distributed/fsdp/fully_sharded_data_parallel.py [hunk 4]
```diff
@@ -776,6 +776,22 @@ def _full_post_state_dict_hook(
...
+        """
+        Hook that runs after model.state_dict() is called before returning result to
+        user. For FSDP, we may have to clone the tensors in state_dict as params go
+        back to sharded version after summon_full_params ends, and also remove
+        "_fsdp_wrapped_module" prefix.
+        """
+        for key in state_dict.keys():
+            # Due to recursive call of summon_full_params, avoid unnecessary reclone of
+            if (
+                self.training_state == TrainingState_.SUMMON_FULL_PARAMS and
```
Comment (fegin) [inline torch/distributed/fsdp/fully_sharded_data_parallel.py:788]: Curious, is there any chance that this hook will be called with `self.training_state` is not `SUMMON_FULL_PARAMS`?

File: torch/distributed/fsdp/fully_sharded_data_parallel.py [hunk 5]
```diff
@@ -776,6 +776,22 @@ def _full_post_state_dict_hook(
...
+                not getattr(state_dict[key], "_has_been_cloned", False)
+            ):
+                state_dict[key] = state_dict[key].clone()
```
Comment (fegin) [inline torch/distributed/fsdp/fully_sharded_data_parallel.py:791]: Is `detach()` needed here?

**Review by fegin — COMMENTED** [review]
LGTM, just a few nits.

File: torch/distributed/fsdp/fully_sharded_data_parallel.py (hunk 5 above)
Comment (rohan-varma) [reply torch/distributed/fsdp/fully_sharded_data_parallel.py:791]: good point, looks like .clone().detach() is the typical pattern for tensor cloning.

File: torch/distributed/fsdp/fully_sharded_data_parallel.py (hunk 4 above)
Comment (rohan-varma) [reply torch/distributed/fsdp/fully_sharded_data_parallel.py:788]: Good point, it should never be called as far as I understand. I'll see if I can actually just change this to an assert.

File: torch/distributed/fsdp/fully_sharded_data_parallel.py [hunk 6]
```diff
@@ -845,8 +862,18 @@ def state_dict(self, *args, **kwargs):
...
+        self._lazy_init()
         if self._state_dict_type == StateDictType.FULL_STATE_DICT:
-            return super().state_dict(*args, **kwargs)
+            if self.training_state != TrainingState_.SUMMON_FULL_PARAMS:
+                with self._summon_full_params(recurse=False, writeback=False):
+                    state_dict = super().state_dict(*args, **kwargs)
+            else:
+                state_dict = super().state_dict(*args, **kwargs)
+
+            # TODO: support offload to CPU in post state dict hook.
```
Comment (rohan-varma) [inline torch/distributed/fsdp/fully_sharded_data_parallel.py:874]: will add the issue

File: torch/distributed/fsdp/fully_sharded_data_parallel.py [hunk 7]
```diff
@@ -845,8 +862,18 @@ def state_dict(self, *args, **kwargs):
...
             torch.cuda.synchronize()
+
+        self._lazy_init()
         if self._state_dict_type == StateDictType.FULL_STATE_DICT:
-            return super().state_dict(*args, **kwargs)
+            if self.training_state != TrainingState_.SUMMON_FULL_PARAMS:
+                with self._summon_full_params(recurse=False, writeback=False):
+                    state_dict = super().state_dict(*args, **kwargs)
+            else:
+                state_dict = super().state_dict(*args, **kwargs)
```
Comment (rohan-varma) [inline torch/distributed/fsdp/fully_sharded_data_parallel.py:872]: need to add a UT where we are already in summon_full_params context

File: torch/distributed/fsdp/fully_sharded_data_parallel.py [hunk 8]
```diff
@@ -865,12 +892,27 @@ def local_state_dict(self, *args: Any, **kwargs: Any) -> Any:
         with self.state_dict_type(StateDictType.LOCAL_STATE_DICT):
             return self.state_dict(*args, **kwargs)
 
+    def full_state_dict(self, *args: Any, **kwargs: Any) -> Any:
```
Comment (zhaojuanmao) [inline torch/distributed/fsdp/fully_sharded_data_parallel.py:895]: synced up offline, we can remove full_state_dict and load_full_state_dict interfaces

**Review by zhaojuanmao — COMMENTED** [review]
looks good, just one minor comment regarding removing full_state_dict() interface as synced offline

**Review by zhaojuanmao — APPROVED** [review]

**Review by fegin — APPROVED** [review]
LGTM
